pkg/networkserver/downlink.go

```diff
@@ -504,7 +504,7 @@ func (ns *NetworkServer) generateDataDownlink(ctx context.Context, dev *ttnpb.En
 	ctx = log.NewContext(ctx, logger)
 
 	if len(cmdBuf) > 0 && (!cmdsInFOpts || macspec.EncryptFOpts(dev.MacState.LorawanVersion)) {
-		if types.MustAES128Key(dev.GetSession().GetKeys().GetNwkSEncKey().GetKey()).OrZero().IsZero() {
+		if dev.GetSession().GetKeys().GetNwkSEncKey() == nil {
```

Comment:
So this check has a pretty funny history.

The old check was `len(dev.Session.GetKeys().GetSNwkSIntKey().GetKey()) == 0` which was always going to be false since `GetKey()` returns `*types.AES128Key`, which is always going to have length 16.

We then changed this to the `IsZero` variant, but this was wrong too, because the session keys are stored in their wrapped form at rest, so `GetKey()` will always return `nil` actually.